code/datums/abilities/wraith.dm

```diff
@@ -147,44 +147,83 @@
 		if (!M && error)
 			boutput(holder.owner, "<span class='alert'>[pick("This body is too decrepit to be of any use.", "This corpse has already been run through the wringer.", "There's nothing useful left.", "This corpse is worthless now.")]</span>")
 			return 1
-		logTheThing("combat", usr, null, "absorbs the corpse of [key_name(M)] as a wraith.")
-
-		//Make the corpse all grody and skeleton-y
-		M.decomp_stage = 4
-		if (M.organHolder && M.organHolder.brain)
-			qdel(M.organHolder.brain)
-		M.set_face_icon_dirty()
-		M.set_body_icon_dirty()
-		particleMaster.SpawnSystem(new /datum/particleSystem/localSmoke("#000000", 5, locate(M.x, M.y, M.z)))
-
-		holder.regenRate *= 2.0
-		holder.owner:onAbsorb(M)
-		//Messages for everyone!
-		boutput(holder.owner, "<span class='alert'><strong>[pick("You draw the essence of death out of [M]'s corpse!", "You drain the last scraps of life out of [M]'s corpse!")]</strong></span>")
-		playsound(M, "sound/voice/wraith/wraithsoulsucc[rand(1, 2)].ogg", 60, 0)
-		for (var/mob/living/V in viewers(7, holder.owner))
-			boutput(V, "<span class='alert'><strong>[pick("Black smoke rises from [M]'s corpse! Freaky!", "[M]'s corpse suddenly rots to nothing but bone in moments!")]</strong></span>")
-
 
+		actions.start(new/datum/action/bar/icon/absorb_corpse(M,src,get_turf(holder.owner)), M)
```

Comment:
Is there any particular reason this is a bar on the corpse instead of the wraith?
Putting it on the wraith would simplify things, as you can just use the `INTERRUPT_MOVE` flag to interrupt on movement. There's also a var somewhere to put the bar on the target instead of the user if it's an aesthetic thing.